src/stan/model/indexing/rvalue_cl.hpp

```diff
@@ -0,0 +1,395 @@
+}
+}  // namespace internal
+
+// prim
+/**
+ * Index a prim kernel generator expression with one index.
+ *
+ * @tparam Expr type of the exoression
+ * @tparam RowIndex type of index
+ * @param expr a prim kernel generator expression to index
+ * @param name string form of expression being evaluated
+ * @param row_index index
+ * @return result of indexing
+ */
+template <typename Expr, typename RowIndex,
+          require_all_kernel_expressions_and_none_scalar_t<Expr>* = nullptr>
+inline auto rvalue(Expr&& expr, const char* name, const RowIndex& row_index) {
+  auto checks = internal::index_check(row_index, name, expr.rows());
+  try {
+    math::index_apply<std::tuple_size<decltype(checks)>::value>(
+        [&checks](auto... Is) {
+          math::results(std::get<Is>(checks).first...)
+              = math::expressions(std::get<Is>(checks).second...);
+        });
+  } catch (const std::domain_error& e) {
+    throw std::out_of_range(e.what());
+  }
+  return math::indexing(expr,
+                        internal::cl_row_index(row_index, expr.rows(), name),
+                        math::col_index(-1, expr.cols()));
+}
+
+/**
+ * Index a prim kernel generator expression with two indices.
+ *
+ * @tparam Expr type of the exoression
+ * @tparam RowIndex type of row index
+ * @tparam ColIndex type of column index
+ * @param expr a prim kernel generator expression to index
+ * @param name string form of expression being evaluated
+ * @param row_index row index
+ * @param col_index column index
+ * @return result of indexing
+ */
+template <typename Expr, typename RowIndex, typename ColIndex,
+          require_all_kernel_expressions_and_none_scalar_t<Expr>* = nullptr,
+          require_any_not_same_t<RowIndex, ColIndex, index_uni>* = nullptr>
+inline auto rvalue(Expr&& expr, const char* name, const RowIndex& row_index,
+                   const ColIndex& col_index) {
+  auto checks1 = internal::index_check(row_index, name, expr.rows());
+  auto checks2 = internal::index_check(col_index, name, expr.cols());
+  try {
+    math::index_apply<std::tuple_size<decltype(checks1)>::value>(
+        [&checks1](auto... Is) {
+          math::results(std::get<Is>(checks1).first...)
+              = math::expressions(std::get<Is>(checks1).second...);
+        });
+    math::index_apply<std::tuple_size<decltype(checks2)>::value>(
+        [&checks2](auto... Is) {
+          math::results(std::get<Is>(checks2).first...)
+              = math::expressions(std::get<Is>(checks2).second...);
+        });
+  } catch (const std::domain_error& e) {
+    throw std::out_of_range(e.what());
+  }
+
+  return math::indexing(expr,
+                        internal::cl_row_index(row_index, expr.rows(), name),
+                        internal::cl_col_index(col_index, expr.cols(), name));
+}
+
+/**
+ * Index a prim kernel generator expression with two single indices.
+ *
+ * @tparam Expr type of the exoression
```

Comment:
do a find/replace for `exoression`

```suggestion
 * @tparam Expr type of the expression
```

Reply:
Fixed.